designs/extensible-cli-and-scaffolding-plugins-phase-2.md

```diff
@@ -0,0 +1,203 @@
+
+* Pros
+  * This provides flexibility for the user to specify the file paths that the plugin would be placed in and Kubebuilder could discover the binaries in those user specified file paths.
+
+  * No constraints on plugin binary naming or directory placements from the Kubebuilder side.
+
+  * Provides a default value for the plugin directory in case user wants to use that to drop their plugins.
+
+* Cons
+  * Adds some complexity as we need to have an extra flag to have the user provide those file paths and/or make PROJECT file changes to add an extra field to `plugins` field that is a comma separated list of file paths for every plugin that is defined in the `layout` field.
+
+#### Approach #2
+
+This is how [kustomize](https://kubectl.docs.kubernetes.io/guides/extending_kustomize/) discovers plugins.
+The PROJECT file in Kubebuilder has `plugins` field and each plugin will have the resources: Group, Kind and Version, so every plugin defined in the PROJECT file configuration can get its own GVK directory for the executable to be placed in and Kubebuilder will search for a plugin with the name of the plugin in the GVK directory of the plugin. Once KB successfully locates the plugin, it will run the plugin using the CLI.
+
+Following `kustomize`'s way of placement, every plugin gets its own directory such as:
+
+```project
+    $XDG_CONFIG_HOME/kubebuilder/plugins
+    /${group}/${version}/LOWERCASE(${kind}) 
+```
+
+The default value of XDG_CONFIG_HOME is $HOME/.config.
+
+* Pros
+  * `kustomize` which is popular and robust tool, follows this approach in which `apiVersion` and `kind` fields are used to locate the plugin.
+
+  * This appoach does not enforce naming of plugins but enforces the directories the plugins have to be in, which is a more common and typical approach.
+
+  * The one-plugin-per-directory requirement eases creation of a plugin bundle for sharing.
+
+* Cons
+  * This makes it mandatory for every plugin that is specified in the `layout` field to have additional configuration information such as Group, Version and Kind in the `plugins` field in the PROJECT file for it to be discoverable. Currently, the `plugins` field is more like an optional configuration storage field.
+
+#### Approach #3
+
+Another approach I would like to mention is adding plugin executables with a prefix `kubebuilder-` followed by the plugin name to the PATH variable. This will enable KB to traverse through the PATH looking for the plugin executables starting with the prefix `kubebuilder-` and matching by the plugin name that was provided in the CLI. Furthermore, a check should be added to verify that the match is an executable or not and return an error if it's not an executable. This approach provides a lot of flexibility in terms of plugin discovery as all the user needs to do is to add the plugin executable to the PATH and Kubebuilder will discover it.
```

Comment:
A doubt - wouldn't this be more error prone in the sense, I can have two versions of same plugin provided in kb command. Since we don't have a way to figure this out based on a file name, we would end up calling the same plugin twice. For example, in the previous approach we can error out immediately when the cli has two plugins in a command with same `group`, `kind`. Eventually the plugin execution will error anyway, but just a scenario where we could have additional validation before allowing the plugins to execute.

Reply:
Good point @varshaprasad96. I agree that the GVK path scheme would be better to handle these scenarios. Hence, I moved Approach 2 and 3 to alternatives section. I repurposed the first approach be a `${pluginname}/${pluginversion}` scheme instead of GVK exactly.